src/libstd/sys/unix/process.rs

```diff
@@ -283,6 +363,24 @@ impl Process {
         if !setup(out_fd, libc::STDOUT_FILENO) { fail(&mut output) }
         if !setup(err_fd, libc::STDERR_FILENO) { fail(&mut output) }
 
+        let close_or_leak = |fd: RawFd| {
+            if cfg.leak_fds_whitelist.contains(&fd) {
+                let flag_orig = libc::fcntl(fd, libc::F_GETFD, 0);
+                if flag_orig < 0 {
+                    return;
+                }
+                let flag_new = flag_orig & !libc::FD_CLOEXEC;
+                if flag_orig != flag_new {
+                    let _ = libc::fcntl(fd, libc::F_SETFD, flag_new);
+                }
+            } else {
+                let _ = libc::close(fd);
+            }
+        };
+        if !cfg.leak_fds {
+            walk_open_fd(close_or_leak);
```

Comment:
You can use the `FIOCLEX` `ioctl` to set the flag atomically and in one system call. I'm not sure what you mean about how no more file descriptors are created because there are no threads. I was talking about how other threads that weren't copied over to the new process by the fork may be holding locks deep inside `malloc` or many other functions and so cause deadlock in the forked process. I'm not sure what you mean about whether or not `getdents` is worth it. If you want to implement this feature correctly you have to use `getdents`. If you cannot implement this feature correctly then you should not implement it.